keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,645 @@
+principle and increases ease of maintenance.
+
+The simplest way to implement this in Kustomize is to create an overlay for
+each component, and have a top-level overlay include only the components that
+it requires. The problem with this approach is that the mechanism that
+Kustomize uses to interpret and ultimately apply patches does not allow for
+mutating different aspects of the very same base resource in parallel without
+modifying its GVKN (unique identifier). That is, sibling overlays that operate
+on the same level of the hierarchy and are not chained (i.e., one does not
+import the other as base) cannot modify their common parent due to resource ID
+conflicts.
+
+For this reason, we need to provide a new type of kustomization that will help
+Kustomize support the composition model, besides the inheritance model.
+
+### Goals
+- Introduce the notion of components in kustomize and, eventually in kubectl
+- Provide the implementation that allows users to define components, i.e.,
+  portable overlays that are able to modify a set of base resources without
+  conflicts, since patches are serialized
+- Maintain existing, stable interfaces that end-users are already familiar
+  with, i.e., offer components as an optional feature
+
+### Non-Goals
+- Overshadow resources with components
+- Abolish inheritance over composition in Kustomize
+
+## Proposal
+
+Create a new kustomization type called `Component`, and a new kustomization
+field called `components`.
+
+A kustomization that is marked as a `Component` has basically the same
+capabilities as a normal kustomization. The main distinction is that they are
+evaluated **after** the resources of the parent kustomization (overlay or
+component) have been accumulated, and **on top** of them. This means that:
+
+* A component with transformers can transform the resources that an overlay has
+  previously specified in the `resources` field. Components with patches do not
+  have to include the target resource in their `resources` field.
+* Multiple components can extend and transform the same set of resources
+  **sequentially**. This is in contrast to overlays, which cannot alter the same
+  base resources, because they clone them and extend them **in parallel**.
+
+In order to create a component, a user can add the following to its
+`kustomization.yaml`:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+```
+
+In order to use a component, the user can refer to it in their
+`kustomization.yaml` via the `components` field:
+
+```yaml:
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+resources:
+  - ../../base
+  - resource1
+  - resource2
+
+components:
+  - ../components/component1
+  - ../components/component2
+```
+
+Note that a component cannot be added to the `resources:` list, and a
+resource/`Kustomization` cannot be added to the `components:` list.
+
+### User Story
+
+Suppose that a company has written a very simple Web application:
```

Comment:
I am not sure I like the "company" wording. Would it make sense to change the whole section in first plural person, i.e., "we"?

"Suppose that a company has written a very simple Web application" can become "Suppose a very simple Web application:"

Reply:
You're right. Since it's a user story, I've changed it to refer to a user, in a third personal plural form (they/them).